sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc

```diff
@@ -0,0 +1,454 @@
+properties)`.
+
+[source,c++]
+----
+template <typename Group, typename Accessor, typename Properties = ext::oneapi::experimental::properties<>>
+void joint_prefetch(Group g, Accessor* acc, size_t offset, Properties properties = {});
+----
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>>` is
+`true`.
+
+_Preconditions_: `acc`, `offset` and `properties` must be the same for all
+work-items in group `g`.
+
+_Effects_: Equivalent to `joint_prefetch(g, (void*) (acc.get_pointer() +
+offset), sizeof(Accessor::value_type), properties)`.
+
+[source,c++]
+----
+template <typename Group, typename Accessor, typename Properties = ext::oneapi::experimental::properties<>>
+void joint_prefetch(Group g, Accessor* acc, size_t offset, size_t count, Properties properties = {});
+----
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>>` is
+`true`.
+
+_Preconditions_: `acc`, `offset`, `count` and `properties` must be the same for
+all work-items in group `g`.
+
+_Effects_: Equivalent to `joint_prefetch(g, (void*) (acc.get_pointer() +
+offset), count * sizeof(Accessor::value_type), properties)`.
+
+
+==== Usage examples
+
+[source,c++]
+----
+using syclex = sycl::ext::oneapi::experimental;
+
+q.parallel_for(sycl::nd_range{N, L}, [=](sycl::nd_item<1> it) {
+  auto sg = it.get_sub_group();
+  for (int j = sg.get_local_id(); j < M; j += sg.get_max_local_range()) {
+    syclex::joint_prefetch(sg, &data[j + 100], sg.get_max_local_range(), syclex::properties{syclex::prefetch_hint_L3});
+    foo(sg, data[j]);
+  }
+});
+----
+
+
+== Issues
+
+. Which level of cache should be targeted for an empty property list?
++
+--
+*UNRESOLVED*:
+Defaulting to the lowest level of cache may be expected by some users, who
+would like the prefetch to place data as close to the compute units as
+possible. Defaulting to the highest level of cache may be expected by other
+users, since that level typically has the highest capacity and may contain
+data from all other levels -- naive usage of prefetches in this case would be
+less likely to cause thrashing across multiple levels of cache.
+
+The current draft of this extension sets the default as the lowest level,
+consistent with the behavior of the `prefetch` pragmas proposed for OpenMP.
+Developers who want to prefetch data into specific levels of cache can simply
+override this behavior, and can prefetch into the last level of cache by
+specifing `prefetch_hint_L4`.
+--
+
+. How should multi-dimensional prefetches be handled?
++
+--
+*UNRESOLVED*:
+Some developers think of multi-dimensional accessors in terms of the underlying
+(linearized) memory, and would expect to describe prefetches in terms of scalar
+offsets and counts. Other developers might expect prefetches using
+multi-dimensional accessors to accept offsets and counts decribed using `id`
+and `range` objects.
+
+The current draft of this extension limits functionality to one-dimensional
+accessors.
```

Comment:
I think there is little precedent for using a linear index into a multi-dimensional accessor.  For example, `operator[]` of accessor cannot take a linearized index.

I think it would be sensible for `prefetch` to take an `id<Dimension>` as the offset parameter.  For a 1-dimensional accessor, the user could still pass `size_t` (I think) because there is an implicit conversion from `size_t` to `id<1>`.

It might also make sense to have an overload of `prefetch` for 0-dimensional accessors, which did not take an offset.  This would be consistent with the member functions on `accessor`.

Reply:
I've extended support to 2- and 3-dimensional accessors in 8f71e4320952.

I think we should hold off on 0-dimensional accessors until somebody has a use-case.  I'm used to seeing prefetches in iterative algorithms (because you need some work to hide the prefetch behind), and I think prefetching the only element in an accessor is going to be a corner-case.  `prefetch(accessor.get_pointer(), 1)` would support this case in the meantime, so the functionality is still there even without an overload.